### PR TITLE
[TorchMhlo] enable matmul uts

### PR DESCRIPTION
**Pull request by fortianyou**

### CODE REVIEW

File: pytorch_blade/src/compiler/mlir/converters/torch_mlir_op_filter.cpp [hunk 1]
```diff
@@ -44,7 +44,9 @@ std::unordered_set<std::string> GetTorchMlirWhiteList() {
       "aten::mul",      "aten::eq",          "aten::lt",
       "aten::ne",       "aten::__and__",     "aten::floor_divide",
       "prim::Constant", "aten::expand_as",   "aten::repeat",
-      "aten::expand",   "aten::sum",         "aten::mean"};
+      "aten::expand",   "aten::sum",         "aten::mean",
```
Comment (wyzero) [inline pytorch_blade/src/compiler/mlir/converters/torch_mlir_op_filter.cpp:47]: Should we sort this list by name? and maybe one line one op is easier to maintain (need to clang-format off to protect)?

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp [hunk 2]
```diff
@@ -20,21 +20,41 @@
...
+
 // Create a 32-bit float constant operator from a float
 Value getMhloConstTensorSingleF32(
     PatternRewriter& rewriter,
     Operation* op,
     float val) {
-  auto const_type = RankedTensorType::get({}, rewriter.getF32Type());
-  auto const_attr = DenseElementsAttr::get(const_type, val);
+  auto constType = RankedTensorType::get({}, rewriter.getF32Type());
+  auto const_attr = DenseElementsAttr::get(constType, val);
```
Comment (wyzero) [inline pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp:54]: using `constAttr` as well?

File: pytorch_blade/src/compiler/mlir/converters/torch_mlir_op_filter.cpp (hunk 1 above)
Comment (fortianyou) [reply pytorch_blade/src/compiler/mlir/converters/torch_mlir_op_filter.cpp:47]: I will resort the list in another PR.

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp [hunk 3]
```diff
@@ -130,44 +150,142 @@ template llvm::Optional<Value> getConstTensor<int64_t>(
...
     ArrayRef<int64_t> shape);
 
+std::vector<Value> getDimSizesOfTensor(
+    PatternRewriter& rewriter,
+    Operation* op,
+    Value value,
+    ArrayRef<int64_t> inpDims) {
+  auto currentKnowledge = ValueKnowledge::getKnowledgeFromType(value.getType());
+  std::vector<Value> dimSizes;
+  if (!currentKnowledge.hasRank) {
```
Comment (wyzero) [inline pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp:160]: nit: maybe return a `llvm::Optional<std::vector<Value>>` is better? I see here return directly while the callee can not know it's a failure case actually.

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp [hunk 4]
```diff
@@ -130,44 +150,142 @@ template llvm::Optional<Value> getConstTensor<int64_t>(
...
-    dim_sizes.emplace_back(rewriter.create<mlir::arith::IndexCastOp>(
+    // if (d_size == ShapedType::kDynamicSize) {
+    dimSizes.emplace_back(rewriter.create<arith::IndexCastOp>(
         loc,
         rewriter.getI32Type(),
         rewriter.create<tensor::DimOp>(loc, value, d)));
     // } else {
-    //   dim_sizes.emplace_back(
-    //     rewriter.create<mlir::arith::ConstantOp>(
+    //   dimSizes.emplace_back(
```
Comment (wyzero) [inline pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp:199]: why should we keep such unused code?

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp [hunk 5]
```diff
@@ -20,21 +20,41 @@
...
+#include <iostream>
 #include "torch-mlir/Conversion/TorchToMhlo/MhloLegalizeUtils.h"
 
 namespace mlir {
 namespace mhlo {
 
+std::vector<int64_t> rangeIndices(int64_t min, int64_t max) {
+  std::vector<int64_t> range;
+  for (int64_t k = min; k < max; ++k) {
+    range.push_back(k);
```
Comment (qiuxiafei) [inline pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp:32]: 1. is this identical to `std::iota`?
2. nit: `range.reserve`

File: pytorch_blade/src/torch-mlir/include/torch-mlir/Conversion/TorchToMhlo/MhloLegalizeUtils.h [hunk 6]
```diff
@@ -124,6 +126,48 @@ llvm::Optional<Value> getMhloShapeOfTensor(
     Value value);
 
 Value getNumelOfTensor(PatternRewriter& rewriter, Operation* op, Value value);
+
+llvm::Optional<Value> getUnsqueezedTensor(
```
Comment (qiuxiafei) [inline pytorch_blade/src/torch-mlir/include/torch-mlir/Conversion/TorchToMhlo/MhloLegalizeUtils.h:130]: nit: `create` seems better than `get`

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp (hunk 2 above)
Comment (qiuxiafei) [reply pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp:54]: I'm completely confused by mixed this coding styling ...

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp (hunk 3 above)
Comment (qiuxiafei) [reply pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp:160]: would empty `dimSizes` get mixed up with scalar type?

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp [hunk 7]
```diff
@@ -182,5 +300,357 @@ Value getNumelOfTensor(PatternRewriter& rewriter, Operation* op, Value value) {
   numel = rewriter.create<tensor::FromElementsOp>(loc, ArrayRef<Value>{numel});
   return numel;
 }
+
+Value getReshapedTensor(
+    PatternRewriter& rewriter,
+    Operation* op,
+    Value tensor,
+    ArrayRef<int64_t> shape,
+    ArrayRef<Value> dimSizes) {
```
Comment (qiuxiafei) [inline pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp:309]: Any comments or doc on these helper function?

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp (hunk 4 above)
Comment (fortianyou) [reply pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp:199]: done

File: pytorch_blade/src/torch-mlir/include/torch-mlir/Conversion/TorchToMhlo/MhloLegalizeUtils.h (hunk 6 above)
Comment (fortianyou) [reply pytorch_blade/src/torch-mlir/include/torch-mlir/Conversion/TorchToMhlo/MhloLegalizeUtils.h:130]: It's named with "get" just to be consistent with other functions.

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp (hunk 3 above)
Comment (fortianyou) [reply pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp:161]: I added codes to emit an error message if they meet illegal usages.

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp (hunk 5 above)
Comment (fortianyou) [reply pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp]: Change to `std::iota`

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp (hunk 7 above)
Comment (fortianyou) [reply pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp:312]: done

**Review by Yancey1989 — APPROVED** [review]
LGTM

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp [hunk 8]
```diff
@@ -20,11 +20,21 @@
...
 #include <mlir-hlo/Dialect/mhlo/IR/hlo_ops.h> // from tf repo
 
+#include <numeric>
 #include "torch-mlir/Conversion/TorchToMhlo/MhloLegalizeUtils.h"
 
 namespace mlir {
 namespace mhlo {
 
+std::vector<int64_t> rangeIndices(int64_t min, int64_t max) {
+  if (min > max)
```
Comment (qiuxiafei) [inline pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/MhloLegalizeUtils.cpp:30]: Assertion or swap silently ?

**Pull request merged**